docs/source/configuration.rst
Configuration
#############

Circus can be configured using an ini-style configuration file.

Example::

    [circus]
    check_delay = 5
    endpoint = tcp://127.0.0.1:5555
    pubsub_endpoint = tcp://127.0.0.1:5556
    include = *.more.config.ini

    [watcher:myprogram]
    cmd = python
    args = -u myprogram.py $(circus.wid) $(ENV.VAR)
    warmup_delay = 0
    numprocesses = 5

    # hook
    hooks.before_start = my.hooks.control_redis

    # will push in test.log the stream every 300 ms
    stdout_stream.class = FileStream
    stdout_stream.filename = test.log

    # optionally rotate the log file when it reaches 1 gb
    # and save 5 copied of rotated files
    stdout_stream.max_bytes = 1073741824
    stdout_stream.backup_count = 5

    [env:myprogram]
    PATH = $PATH:/bin
    CAKE = lie

    [plugin:statsd]
    use = circus.plugins.statsd.StatsdEmitter
    host = localhost
    port = 8125
    sample_rate = 1.0
    application_name = example

    [socket:web]
    host = localhost
    port = 8080


circus - single section
=======================
    **endpoint**
        The ZMQ socket used to manage Circus via **circusctl**.
        (default: *tcp://127.0.0.1:5555*)
    **pubsub_endpoint**
        The ZMQ PUB/SUB socket receiving publications of events.
        (default: *tcp://127.0.0.1:5556*)
    **statsd**
        If set to True, Circus runs the circusd-stats daemon. (default: False)
    **stats_endpoint**
        The ZMQ PUB/SUB socket receiving publications of stats.
        (default: *tcp://127.0.0.1:5557*)
    **statsd_close_outputs**
        If True sends the circusd-stats stdout/stderr to /dev/null.
        (default: False)
    **check_delay**
        The polling interval in seconds for the ZMQ socket. (default: 5)
    **include**
        List of config files to include. You can use wildcards
        (`*`) to include particular schemes for your files. The paths are
        absolute or relative to the config file. (default: None)
    **include_dir**
        List of config directories. All files matching `*.ini` under each
        directory will be included. The paths are absolute or relative to the
        config file. (default: None)
    **stream_backend**
        Defines the type of backend to use for the streaming. Possible
        values are **thread** or **gevent**. (default: thread)
    **warmup_delay**
        The interval in seconds between two watchers start. Must be an int. (default: 0)
    **httpd**
        If set to True, Circus runs the circushttpd daemon. (default: False)
    **httpd_host**
        The host ran by the circushttpd daemon. (default: localhost)
    **httpd_port**
        The port ran by the circushttpd daemon. (default: 8080)
    **httpd_close_outputs**
        If True, sends the circushttpd stdout/stderr to /dev/null.
        (default: False)
    **debug**
        If set to True, all Circus stout/stderr daemons are redirected to circusd
        stdout/stderr (default: False)
    **pidfile**
        The file that must be used to keep the daemon pid.
    **loglevel**
        The loglevel that we want to see (default: INFO)
    **logoutput**
        The logoutput file where we want to log (default: stdout)


watcher:NAME - as many sections as you want
===========================================
    **NAME**
        The name of the watcher. This name is used in **circusctl**
    **cmd**
        The executable program to run.
    **args**
        Command-line arguments to pass to the program. You can use the python
        format syntax here to build the parameters. Environment variables are
        available, as well as the worker id and the environment variables that
        you passed, if any, with the "env" parameter. See
        :ref:`formatting_cmd` for more information on this.
    **shell**
        If True, the processes are run in the shell (default: False)
    **working_dir**
        The working dir for the processes (default: None)
    **uid**
        The user id or name the command should run with.
        (The current uid is the default).
    **gid**
        The group id or name the command should run
        with. (The current gid is the default).
    **copy_env**
        If set to true, the local environment variables will be copied and
        passed to the workers when spawning them. (Default: False)
    **copy_path**
        If set to true, **sys.path** is passed in the subprocess environ
        using *PYTHONPATH*. **copy_env** has to be true.
        (Default: False)
    **warmup_delay**
        The delay (in seconds) between running processes.
    **autostart**
        If set to false, the watcher will not be started automatically
        when the arbiter starts. The watcher can be started explicitly
        (example: `circusctrl start myprogram`). (Default: True)
    **numprocesses**
        The number of processes to run for this watcher.
    **rlimit_LIMIT**
        Set resource limit LIMIT for the watched processes. The
        config name should match the RLIMIT_* constants (not case
        sensitive) listed in the `Python resource module reference
        <http://docs.python.org/library/resource.html#resource-limits>`_.
        For example, the config line 'rlimit_nofile = 500' sets the maximum
        number of open files to 500.
    **stderr_stream.class**
        A fully qualified Python class name that will be instanciated, and
        will receive the **stderr** stream of all processes in its
        :func:`__call__` method.

        Circus provides some stream classes you can use without prefix:

        - :class:`FileStream`: writes in a file
        - :class:`QueueStream`: write in a memory Queue
        - :class:`StdoutStream`: writes in the stdout
        - :class:`FancyStdoutStream`: writes colored output with time prefixes in the stdout

    **stderr_stream.***
        All options starting with *stderr_stream.* other than *class* will
        be passed the constructor when creating an instance of the
        class defined in **stderr_stream.class**.
    **stdout_stream.class**
        A fully qualified Python class name that will be instanciated, and
        will receive the **stdout** stream of all processes in its
        :func:`__call__` method.

        Circus provides soem stream classes you can use without prefix:

        - :class:`FileStream`: writes in a file
        - :class:`QueueStream`: write in a memory Queue
        - :class:`StdoutStream`: writes in the stdout
        - :class:`FancyStdoutStream`: writes colored output with time prefixes in the stdout

    **stdout_stream.***
        All options starting with *stdout_stream.* other than *class* will
        be passed the constructor when creating an instance of the
        class defined in **stdout_stream.class**.

    **close_child_stdout**
        If set to True, the sdout stream of each process will be sent to
        /dev/null after the fork. Defaults to False.

    **close_child_stderr**
        If set to True, the sderr stream of each process will be sent to
        /dev/null after the fork. Defaults to False.

    **send_hup**
        If True, a process reload will be done by sending the SIGHUP signal.
        Defaults to False.

    **stop_signal**
        The signal to send when stopping the process. Can be specified as a
        number or a signal name. Signal names are case-insensitive and can
        include 'SIG' or not. So valid examples include `quit`, `INT`,
        `SIGTERM` and `3`.
        Defaults to SIGTERM.

    **max_retry**
        The number of times we attempt to start a process, before
        we abandon and stop the whole watcher. Defaults to 5.
        Set to -1 to disable max_retry and retry indefinitely.

.. _graceful_timeout:

    **graceful_timeout**
        The number of seconds to wait for a process to terminate gracefully
        before killing it.

        When stopping a process, we first send it a :ref:`stop_signal`. A worker
        may catch this signal to perform clean up operations before exiting.
        If the worker is still active after graceful_timeout seconds, we send
        it a SIGKILL signal.  It is not possible to catch SIGKILL signals so
        the worker will stop.

    **priority**
        Integer that defines a priority for the watcher. When the
        Arbiter do some operations on all watchers, it will sort them
        with this field, from the bigger number to the smallest.
        Defaults to 0.

    **singleton**
        If set to True, this watcher will have at the most one process.
        Defaults to False.

    **use_sockets**
        If set to True, this watcher will be able to access defined sockets
        via their file descriptors. If False, all parent fds are closed
        when the child process is forked. Defaults to False.

    **max_age**
        If set then the process will be restarted sometime after max_age
        seconds. This is useful when processes deal with pool of connectors:
        restarting processes improves the load balancing. Defaults to being
        disabled.

    **max_age_variance**
        If max_age is set then the process will live between max_age and
        max_age + random(0, max_age_variance) seconds. This avoids restarting
        all processes for a watcher at once. Defaults to 30 seconds.

    **on_demand**
        If set to True, the processes will be started only after the first
        connection to one of the configured sockets (see below). If a restart
        is needed, it will be only triggered at the next socket event.

    **hooks.***
        Available hooks: **before_start**, **before_spawn**, **after_start**,
        **before_stop**, **after_stop**, **before_signal**, **after_signal**

        Define callback functions that hook into the watcher startup/shutdown process.

        If the hook returns **False** and if the hook is one of
        **before_start**, **before_spawn** or  **after_start**, the startup
        will be aborted.

        If the hook is **before_signal** and returns **False**, then the
        corresponding signal will not be sent (except SIGKILL which is always
        sent)

        Notice that a hook that fails during the stopping process will not
        abort it.

        The callback definition can be followed by a boolean flag separated by a
        comma. When the flag is set to **true**, any error occuring in the
        hook will be ignored. If set to **false** (the default), the hook
        will return **False**.

        More on :ref:`hooks`.

    **virtualenv**
        When provided, points to the root of a Virtualenv directory. The
        watcher will scan the local **site-packages** and loads its content
        into the execution environment. Must be used with **copy_env** set
        to True. Defaults to None.

    **respawn**
        If set to False, the processes handled by a watcher will not be
        respawned automatically. (default: True)



socket:NAME - as many sections as you want
==========================================
    **host**
        The host of the socket. Defaults to 'localhost'
    **port**
        The port. Defaults to 8080.
    **family**
        The socket family. Can be 'AF_UNIX', 'AF_INET' or 'AF_INET6'.
        Defaults to 'AF_INET'.
    **type**
        The socket type. Can be 'SOCK_STREAM', 'SOCK_DGRAM', 'SOCK_RAW',
        'SOCK_RDM' or 'SOCK_SEQPACKET'. Defaults to 'SOCK_STREAM'.
    **interface**
        When provided a network interface name like 'eth0', binds the socket
        to that particular device so that only packets received from that
        particular interface are processed by the socket.
        This can be used for example to limit which device to bind when
        binding on IN_ADDR_ANY (0.0.0.0) or IN_ADDR_BROADCAST
        (255.255.255.255). Note that this only works for some socket types,
        particularly AF_INET sockets.
    **path**
        When provided a path to a file that will be used as a unix socket
        file. If a path is provided, **family** is forced to AF_UNIX and
        **host** and **port** are ignored.
    **umask**
        When provided, sets the umask that will be used to create an
        AF_UNIX socket. For example, `umask=000` will produce a socket with
        permission `777`.


Once a socket is created, the *${circus.sockets.NAME}* string can be used in the
command (`cmd` or `args`) of a watcher. Circus will replace it by the FD value. The watcher must also
have `use_sockets` set to `True` otherwise the socket will have been closed and
you will get errors when the watcher tries to use it.

Example::

    [watcher:webworker]
    cmd = chaussette --fd $(circus.sockets.webapp) chaussette.util.bench_app
    use_sockets = True

    [socket:webapp]
    host = 127.0.0.1
    port = 8888


plugin:NAME - as many sections as you want
==========================================
    **use**
        The fully qualified name that points to the plugin class.
    **anything else**
        Every other key found in the section is passed to the
        plugin constructor in the **config** mapping.

        You can use all the watcher options, since a plugin is
        started like a watcher.

Circus comes with a few pre-shipped :ref:`plugins <plugins>` but you can also extend them easily by :ref:`developing your own <develop_plugins>`.


env or env[:WATCHERS] - as many sections as you want
====================================================
    **anything**
        The name of an environment variable to assign value to.
        bash style environment substitutions are supported.
        for example, append /bin to `PATH` 'PATH = $PATH:/bin'

Section responsible for delivering environment variable to run processes.

Example::

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env]
    CAKE = lie

The variable `CAKE` will propagated to all watchers defined in config file.

WATCHERS can be a comma separated list of watcher sections to apply this environment to.
if multiple env sections match a watcher, they will be combine in the order they appear in the configuration file.
later entries will take precedence.

Example::

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env:worker1,worker2]
    PATH = /bin

    [env:worker1]
    PATH = $PATH

    [env:worker2]
    CAKE = lie

`worker1` will be run with PATH = $PATH (expanded from the environment circusd was run in)
`worker2` will be run with PATH = /bin and CAKE = lie

It's possible to use wildcards as well.

Example::

    [watcher:worker1]
    cmd = ping 127.0.0.1

    [watcher:worker2]
    cmd = ping 127.0.0.1

    [env:worker*]
    PATH = /bin


Both `worker1` and `worker2` will be run with PATH = /bin


Using environment variables
===========================

When writing your configuration file, you can use environment
variables defined in the *env* section or in *os.environ* itself.

You just have to use the *circus.env.* prefix.

Example::

    [watcher:worker1]
    cmd = $(circus.env.shell)

    [watcher:worker2]
    baz = $(circus.env.user)
    bar = $(circus.env.yeah)
    sup = $(circus.env.oh)

    [socket:socket1]
    port = $(circus.env.port)

    [plugin:plugin1]
    use = some.path
    parameter1 = $(circus.env.plugin_param)

    [env]
    yeah = boo

    [env:worker2]
    oh = ok

If a variable is defined in several places, the most specialized
value has precedence: a variable defined in *env:XXX* will override
a variable defined in *env*, which will override a variable
defined in *os.environ*.

environment substitutions can be used in any section of the configuration
in any section variable.


.. _formatting_cmd:

Formatting the commands and arguments with dynamic variables
============================================================

As you may have seen, it is possible to pass some information that are computed
dynamically when running the processes. Among other things, you can get the
worker id (WID) and all the options that are passed to the :class:`Process`.
Additionally, it is possible to access the options passed to the
:class:`Watcher` which instanciated the process.

.. note::

   The worker id is different from the process id. It's a unique value,
   starting at 1, which is only unique for the watcher.

For instance, if you want to access some variables that are contained in the
environment, you would need to do it with a setting like this::

    cmd = "make-me-a-coffee --sugar $(CIRCUS.ENV.SUGAR_AMOUNT)"

This works with both `cmd` and `args`.

**Important**:

- All variables are prefixed with `circus.`
- The replacement is case insensitive.

Stream configuration
====================

Simple stream class like `QueueStream` and `StdoutStream` don't have
specific attributes but some other stream class may have some:


FileStream
::::::::::

    **filename**
        The file path where log will be written.

    **time_format**
        The strftime format that will be used to prefix each time with a timestamp.
        By default they will be not prefixed.

        i.e: %Y-%m-%d %H:%M:%S

    **max_bytes**
        The max size of the log file before a new file is started.
        If not provided, the file is not rolled over.

    **backup_count**
        The number of log files that will be kept
        By default backup_count is null.


.. note::

    Rollover occurs whenever the current log file is nearly max_bytes in
    length. If backup_count is >= 1, the system will successively create
    new files with the same pathname as the base file, but with extensions
    ".1", ".2" etc. appended to it. For example, with a backup_count of 5
    and a base file name of "app.log", you would get "app.log",
    "app.log.1", "app.log.2", ... through to "app.log.5". The file being
    written to is always "app.log" - when it gets filled up, it is closed
    and renamed to "app.log.1", and if files "app.log.1", "app.log.2" etc.
    exist, then they are renamed to "app.log.2", "app.log.3" etc.
    respectively.

Example::

    [watcher:myprogram]
    cmd = python -m myapp.server

    stdout_stream.class = FileStream
    stdout_stream.filename = test.log
    stdout_stream.time_format = %Y-%m-%d %H:%M:%S
    stdout_stream.max_bytes = 1073741824
    stdout_stream.backup_count = 5


FancyStdoutStram
::::::::::::::::

    **color**
        The name of an ascii color:
            - red
            - green
            - yellow
            - blue
            - magenta
            - cyan
            - white

    **time_format**
        The strftime format that each line will be prefixed with.

        Default to: %Y-%m-%d %H:%M:%S

Example::

    [watcher:myprogram]
    cmd = python -m myapp.server
    stdout_stream.class = FancyStdoutStream
    stdout_stream.color = green
    stdout_stream.time_format = %Y/%m/%d | %H:%M:%S


CHANGES.rst
CHANGES
=======

0.7
---
???


0.6 - 2012-12-18
----------------


* Patching protocols name for sockets - #248
* Don't autoscale graphs. #240
* circusctl: add per command help, from docstrings #217
* Added workers hooks
* Added Debian package - #227
* Added Redis, HTTP Observer, Full stats & Resource plugins
* Now processes can have titles
* Added autocompletion
* Added process/watcher age in the webui
* Added SSH tunnel support
* Now using pyzmq.green
* Added upstart script & Varnish doc
* Added environment variables & sections
* Added unix sockets support
* Added the *respawn* option to have single-run watchers
* Now using tox in the tests
* Allow socket substitution in args
* New doc theme
* New rotation options for streams: max_bytes/backup_count


0.5.2 - 2012-07-26
------------------

* now patching the thread module from the stdlib
  to avoid some Python bugs - #203
* better looking circusctl help screen
* uses pustil get_nice() when available (nice was deprecated) - #208
* added max_age support - #221
* only call listen() on SOCK_STREAM or SOCK_SEQPACKET sockets
* make sure the controller empties the plugins list in update_watchers() - #220
* added --log-level and --log-output to circushttpd
* fix the process killing via the web UI - #219
* now circus is zc.buildout compatible for scripts.
* cleanup the websocket when the client disconnect - #225
* fixed the default value for the endpoint - #199
* splitted circushttpd in logical modules


0.5.1 - 2012-07-11
------------------

* Fixed a bunch of typos in the documentation
* Added the debug option
* Package web-requirements.txt properly
* Added a errno error code in the messages - fixes #111

0.5 - 2012-07-06
----------------

* added socket support
* added a listsocket command
* sockets have stats too !
* fixed a lot of small bugs
* removed the wid - now using pid everywhere
* faster tests
* changed the variables syntax
* use pyzmq's ioloop in more places
* now using iowait for all select() calls
* incr/decr commands now have an nbprocess parameter
* Add a reproduce_env option to watchers
* Add a new UNEXISTING status to the processes
* Added the global *httpd* option to run circushttpd as a watcher


0.4 - 2012-06-12
----------------

* Added a plugin system
* Added a "singleton" option for watchers
* Fixed circus-top screen flickering
* Removed threads from circus.stats in favor of zmq periodic callbacks
* Enhanced the documentation
* Circus client now have a send_message api
* The flapping feature is now a plugin
* Every command line tool have a --version option
* Added a statsd plugin (sends the events from circus to statsd)
* The web UI now uses websockets (via socketio) to get the stats
* The web UI now uses sessions for "flash messages" in the web ui

0.3.4 - 2012-05-30
------------------

- Fixed a race condition that prevented the controller
  to cleanly reap finished processes.
- Now check_flapping can be controlled in the configuration.
  And activated/deactivated per watcher.


0.3.3 - 2012-05-29
------------------

- Fixed the regression on the uid handling

0.3.2 - 2012-05-24
------------------

- allows optional args property to add_watcher command.
- added circushttpd, circus-top and circusd-stats
- allowing Arbiter.add_watcher() to set all Watcher option
- make sure the redirectors are re-created on restarts


0.3.1 - 2012-04-18
------------------

- fix: make sure watcher' defaults aren't overrided
- added a StdoutStream class.

0.3 - 2012-04-18
----------------

- added the streaming feature
- now displaying coverage in the Sphinx doc
- fixed the way the processes are killed (no more SIGQUIT)
- the configuration has been factored out
- setproctitle support


0.2 - 2012-04-04
----------------

- Removed the *show* name. replaced by *watcher*.
- Added support for setting process **rlimit**.
- Added support for include dirs in the config file.
- Fixed a couple of leaking file descriptors.
- Fixed a core dump in the flapping
- Doc improvments
- Make sure circusd errors properly when another circusd
  is running on the same socket.
- get_arbiter now accepts several watchers.
- Fixed the cmd vs args vs executable in the process init.
- Fixed --start on circusctl add


0.1 - 2012-03-20
----------------

- initial release


circus/config.py
import glob
import os
import signal
import warnings
from fnmatch import fnmatch

from circus import logger
from circus.util import (DEFAULT_ENDPOINT_DEALER, DEFAULT_ENDPOINT_SUB,
                         DEFAULT_ENDPOINT_MULTICAST, DEFAULT_ENDPOINT_STATS,
                         StrictConfigParser, replace_gnu_args, to_signum)


def watcher_defaults():
    return {
        'name': '',
        'cmd': '',
        'args': '',
        'numprocesses': 1,
        'warmup_delay': 0,
        'executable': None,
        'working_dir': None,
        'shell': False,
        'uid': None,
        'gid': None,
        'send_hup': False,
        'stop_signal': signal.SIGTERM,
        'stop_children': False,
        'max_retry': 5,
        'graceful_timeout': 30,
        'rlimits': dict(),
        'stderr_stream': dict(),
        'stdout_stream': dict(),
        'priority': 0,
        'use_sockets': False,
        'singleton': False,
        'copy_env': False,
        'copy_path': False,
        'hooks': dict(),
        'respawn': True,
        'autostart': True}


_BOOL_STATES = {'1': True, 'yes': True, 'true': True, 'on': True,
                '0': False, 'no': False, 'false': False, 'off': False}


def to_boolean(value):
    value = value.lower().strip()
    if value not in _BOOL_STATES:
        raise ValueError(value)
    return _BOOL_STATES[value]


class DefaultConfigParser(StrictConfigParser):

    def __init__(self, *args, **kw):
        StrictConfigParser.__init__(self, *args, **kw)
        self._env = dict(os.environ)

    def set_env(self, env):
        self._env = dict(env)

    def get(self, section, option):
        res = StrictConfigParser.get(self, section, option)
        return replace_gnu_args(res, env=self._env)

    def items(self, section, noreplace=False):
        items = StrictConfigParser.items(self, section)
        if noreplace:
            return items

        return [(key, replace_gnu_args(value, env=self._env))
                for key, value in items]

    def dget(self, section, option, default=None, type=str):
        if not self.has_option(section, option):
            return default

        value = self.get(section, option)

        if type is int:
            value = int(value)
        elif type is bool:
            value = self.toboolean(value)
        elif type is not str:
            raise NotImplementedError()

        return value


def read_config(config_path):
    cfg = DefaultConfigParser()
    with open(config_path) as f:
        if hasattr(cfg, 'read_file'):
            cfg.read_file(f)
        else:
            cfg.readfp(f)

    current_dir = os.path.dirname(config_path)

    # load included config files
    includes = []

    def _scan(filename, includes):
        if os.path.abspath(filename) != filename:
            filename = os.path.join(current_dir, filename)

        paths = glob.glob(filename)
        if paths == []:
            logger.warn('%r does not lead to any config. Make sure '
                        'include paths are relative to the main config '
                        'file' % filename)
        includes += paths

    for include_file in cfg.dget('circus', 'include', '').split():
        _scan(include_file, includes)

    for include_dir in cfg.dget('circus', 'include_dir', '').split():
        _scan(os.path.join(include_dir, '*.ini'), includes)

    logger.debug('Reading config files: %s' % includes)
    return cfg, [config_path] + cfg.read(includes)


def get_config(config_file):
    if not os.path.exists(config_file):
        raise IOError("the configuration file %r does not exist\n" %
                      config_file)

    cfg, cfg_files_read = read_config(config_file)
    dget = cfg.dget
    config = {}

    # reading the global environ first
    def _upper(items):
        return [(key.upper(), value) for key, value in items]

    global_env = dict(_upper(os.environ.items()))
    local_env = dict()

    # update environments with [env] section
    if 'env' in cfg.sections():
        local_env.update(dict(_upper(cfg.items('env'))))
        global_env.update(local_env)

    # always set the cfg environment
    cfg.set_env(global_env)

    # main circus options
    config['check'] = dget('circus', 'check_delay', 5, int)
    config['endpoint'] = dget('circus', 'endpoint', DEFAULT_ENDPOINT_DEALER)
    config['pubsub_endpoint'] = dget('circus', 'pubsub_endpoint',
                                     DEFAULT_ENDPOINT_SUB)
    config['multicast_endpoint'] = dget('circus', 'multicast_endpoint',
                                        DEFAULT_ENDPOINT_MULTICAST)
    config['stats_endpoint'] = dget('circus', 'stats_endpoint', None)
    config['statsd'] = dget('circus', 'statsd', False, bool)

    if config['stats_endpoint'] is None:
        config['stats_endpoint'] = DEFAULT_ENDPOINT_STATS
    elif not config['statsd']:
        warnings.warn("You defined a stats_endpoint without "
                      "setting up statsd to True.",
                      DeprecationWarning)
        config['statsd'] = True

    config['warmup_delay'] = dget('circus', 'warmup_delay', 0, int)
    config['httpd'] = dget('circus', 'httpd', False, bool)
    config['httpd_host'] = dget('circus', 'httpd_host', 'localhost', str)
    config['httpd_port'] = dget('circus', 'httpd_port', 8080, int)
    config['debug'] = dget('circus', 'debug', False, bool)
    config['pidfile'] = dget('circus', 'pidfile')
    config['loglevel'] = dget('circus', 'loglevel')
    config['logoutput'] = dget('circus', 'logoutput')
    config['fqdn_prefix'] = dget('circus', 'fqdn_prefix', None, str)

    # Initialize watchers, plugins & sockets to manage
    watchers = []
    plugins = []
    sockets = []

    for section in cfg.sections():
        if section.startswith("socket:"):
            sock = dict(cfg.items(section))
            sock['name'] = section.split("socket:")[-1].lower()
            sockets.append(sock)

        if section.startswith("plugin:"):
            plugin = dict(cfg.items(section))
            plugin['name'] = section
            plugins.append(plugin)

        if section.startswith("watcher:"):
            watcher = watcher_defaults()
            watcher['name'] = section.split("watcher:", 1)[1]

            # create watcher options
            for opt, val in cfg.items(section, noreplace=True):
                if opt == 'cmd':
                    watcher['cmd'] = val
                elif opt == 'args':
                    watcher['args'] = val
                elif opt == 'numprocesses':
                    watcher['numprocesses'] = dget(section, 'numprocesses', 1,
                                                   int)
                elif opt == 'warmup_delay':
                    watcher['warmup_delay'] = dget(section, 'warmup_delay', 0,
                                                   int)
                elif opt == 'executable':
                    watcher['executable'] = dget(section, 'executable', None,
                                                 str)
                elif opt == 'working_dir':
                    watcher['working_dir'] = val
                elif opt == 'shell':
                    watcher['shell'] = dget(section, 'shell', False, bool)
                elif opt == 'uid':
                    watcher['uid'] = val
                elif opt == 'gid':
                    watcher['gid'] = val
                elif opt == 'send_hup':
                    watcher['send_hup'] = dget(section, 'send_hup', False,
                                               bool)
                elif opt == 'stop_signal':
                    watcher['stop_signal'] = to_signum(val)
                elif opt == 'stop_children':
                    watcher['stop_children'] = dget(section, 'stop_children',
                                                    False, bool)
                elif opt == 'check_flapping':
                    watcher['check_flapping'] = dget(section, 'check_flapping',
                                                     True, bool)
                elif opt == 'max_retry':
                    watcher['max_retry'] = dget(section, "max_retry", 5, int)
                elif opt == 'graceful_timeout':
                    watcher['graceful_timeout'] = dget(
                        section, "graceful_timeout", 30, int)
                elif opt.startswith('stderr_stream') or \
                        opt.startswith('stdout_stream'):
                    stream_name, stream_opt = opt.split(".", 1)
                    watcher[stream_name][stream_opt] = val
                elif opt.startswith('rlimit_'):
                    limit = opt[7:]
                    watcher['rlimits'][limit] = int(val)
                elif opt == 'priority':
                    watcher['priority'] = dget(section, "priority", 0, int)
                elif opt == 'use_sockets':
                    watcher['use_sockets'] = dget(section, "use_sockets",
                                                  False, bool)
                elif opt == 'singleton':
                    watcher['singleton'] = dget(section, "singleton", False,
                                                bool)
                elif opt == 'copy_env':
                    watcher['copy_env'] = dget(section, "copy_env", False,
                                               bool)
                elif opt == 'copy_path':
                    watcher['copy_path'] = dget(section, "copy_path", False,
                                                bool)
                elif opt.startswith('hooks.'):
                    hook_name = opt[len('hooks.'):]
                    val = [elmt.strip() for elmt in val.split(',', 1)]
                    if len(val) == 1:
                        val.append(False)
                    else:
                        val[1] = to_boolean(val[1])

                    watcher['hooks'][hook_name] = val

                elif opt == 'respawn':
                    watcher['respawn'] = dget(section, "respawn", True, bool)

                elif opt == 'autostart':
                    watcher['autostart'] = dget(section, "autostart", True,
                                                bool)
                elif opt == 'close_child_stdout':
                    watcher['close_child_stdout'] = dget(section,
                                                         "close_child_stdout",
                                                         False, bool)
                elif opt == 'close_child_stderr':
                    watcher['close_child_stderr'] = dget(section,
                                                         "close_child_stderr",
                                                         False, bool)
                else:
                    # freeform
                    watcher[opt] = val

            if watcher['copy_env']:
                watcher['env'] = dict(global_env)
            else:
                watcher['env'] = dict(local_env)

            watchers.append(watcher)

    # Second pass to make sure env sections apply to all watchers.

    def _extend(target, source):
        for name, value in source:
            if name in target:
                continue
            target[name] = value

    def _expand_vars(target, key, env):
        if isinstance(target[key], str):
            target[key] = replace_gnu_args(target[key], env=env)
        elif isinstance(target[key], dict):
            for k in target[key].keys():
                _expand_vars(target[key], k, env)

    def _expand_section(section, env, exclude=None):
        if exclude is None:
            exclude = ('name', 'env')

        for option in section.keys():
            if option in exclude:
                continue
            _expand_vars(section, option, env)

    # build environment for watcher sections
    for section in cfg.sections():
        if section.startswith('env:'):
            section_elements = section.split("env:", 1)[1]
            watcher_patterns = [s.strip() for s in section_elements.split(',')]
            env_items = dict(_upper(cfg.items(section, noreplace=True)))

            for pattern in watcher_patterns:
                match = [w for w in watchers if fnmatch(w['name'], pattern)]

                for watcher in match:
                    watcher['env'].update(env_items)

    # expand environment for watcher sections
    for watcher in watchers:
        env = dict(global_env)
        env.update(watcher['env'])
        _expand_section(watcher, env)

    config['watchers'] = watchers
    config['plugins'] = plugins
    config['sockets'] = sockets
    return config
